File: library/banner/src/main/java/com/example/banner/holder/DefaultImageHolderView.java
package com.example.banner.holder;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup.LayoutParams;
import android.widget.ImageView;

import androidx.annotation.NonNull;

/**
 * @author fqxyi
 * @date 2017/9/1
 * 本地、网络图片加载
 */
public class DefaultImageHolderView<T> implements Holder<T> {

    private int bannerHeight;
    private int bannerBgColor;

    private ImageView.ScaleType scaleType = ImageView.ScaleType.CENTER_CROP;

    // 显示图片事件
    private ImageShowListener<T> imageShowListener;

    public interface ImageShowListener<T> {
        void showImage(ImageView imageView, T data);
    }

    public void setImageShowListener(ImageShowListener<T> imageShowListener) {
        this.imageShowListener = imageShowListener;
    }

    // 点击事件
    private ImageClickListener<T> clickListener;

    public interface ImageClickListener<T> {
        void click(View view, int position, T data);
    }

    public void setClickListener(ImageClickListener<T> clickListener) {
        this.clickListener = clickListener;
    }

    // 长按事件
    private ImageLongClickListener<T> longListener;

    public interface ImageLongClickListener<T> {
        void longClick(View view, int position, T data);
    }

    public void setLongClickListener(ImageLongClickListener<T> longListener) {
        this.longListener = longListener;
    }

    public DefaultImageHolderView() {
        this.bannerHeight = LayoutParams.MATCH_PARENT;
    }

    public DefaultImageHolderView(int bannerHeight) {
        this.bannerHeight = bannerHeight;
    }

    /**
     * 你可以通过layout文件来创建，也可以像我一样用代码创建，不一定是Image，任何控件都可以进行翻页
     */
    @Override
    public View createView(Context context) {
        ImageView imageView = new ImageView(context);
        LayoutParams params = new LayoutParams(LayoutParams.MATCH_PARENT, bannerHeight);
        imageView.setLayoutParams(params);
        imageView.setScaleType(scaleType);

        if (bannerBgColor != 0) {
            imageView.setBackgroundColor(bannerBgColor);
        }
        return imageView;
    }

    @Override
    public void updateUI(Context context, final View view, final int position, @NonNull final T data) {
        if (view instanceof ImageView && imageShowListener != null) {
            imageShowListener.showImage((ImageView) view, data);
        }
        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (null != clickListener) {
                    clickListener.click(view, position, data);
                }
            }
        });
        view.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                if (null != longListener) {
                    longListener.longClick(view, position, data);
                }
                return false;
            }
        });
    }

    public void setBannerHeight(int bannerHeight) {
        this.bannerHeight = bannerHeight;
    }

    public void setScaleType(ImageView.ScaleType scaleType) {
        this.scaleType = scaleType;
    }

    public void setBannerBgColor(int color) {
        this.bannerBgColor = color;
    }
}

File: library/banner/src/main/java/com/example/banner/holder/DefaultImageHolderView.kt
package com.example.banner.holder

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ImageView.ScaleType

/**
 * @author fqxyi
 * @date 2017/9/1
 * 本地、网络图片加载
 */
class DefaultImageHolderView<T> constructor(private var bannerHeight: Int = ViewGroup.LayoutParams.MATCH_PARENT) : Holder<T> {

    private var bannerBgColor = 0

    private var scaleType = ScaleType.CENTER_CROP

    // 显示图片事件
    private var imageShowListener: ImageShowListener<T>? = null

    interface ImageShowListener<T> {
        fun showImage(imageView: ImageView, data: T)
    }

    fun setImageShowListener(imageShowListener: ImageShowListener<T>?) {
        this.imageShowListener = imageShowListener
    }

    // 点击事件
    private var clickListener: ImageClickListener<T>? = null

    interface ImageClickListener<T> {
        fun click(view: View, position: Int, data: T)
    }

    fun setClickListener(clickListener: ImageClickListener<T>?) {
        this.clickListener = clickListener
    }

    // 长按事件
    private var longListener: ImageLongClickListener<T>? = null

    interface ImageLongClickListener<T> {
        fun longClick(view: View, position: Int, data: T)
    }

    fun setLongClickListener(longListener: ImageLongClickListener<T>?) {
        this.longListener = longListener
    }

    /**
     * 你可以通过layout文件来创建，也可以像我一样用代码创建，不一定是Image，任何控件都可以进行翻页
     */
    override fun createView(context: Context?): View {
        val imageView = ImageView(context)
        val params = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, bannerHeight)
        imageView.layoutParams = params
        imageView.scaleType = scaleType
        if (bannerBgColor != 0) {
            imageView.setBackgroundColor(bannerBgColor)
        }
        return imageView
    }

    override fun updateUI(context: Context?, view: View, position: Int, data: T) {
        if (view is ImageView) {
            imageShowListener?.showImage(view, data)
        }
        view.setOnClickListener {
            clickListener?.click(view, position, data)
        }
        view.setOnLongClickListener {
            longListener?.longClick(view, position, data)
            false
        }
    }

    fun setBannerHeight(bannerHeight: Int) {
        this.bannerHeight = bannerHeight
    }

    fun setScaleType(scaleType: ScaleType) {
        this.scaleType = scaleType
    }

    fun setBannerBgColor(color: Int) {
        bannerBgColor = color
    }
}